velox/functions/sparksql/String.h

```diff
@@ -638,4 +638,153 @@ struct LeftFunction {
+///   corresponding to the character in ``match``. The translation will
+///   happen when any character in ``string`` matching with a character in
+///   ``match``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  folly::F14FastMap<std::string, std::string> unicodeDictionary_;
+  folly::F14FastMap<char, char> asciiDictionary_;
+
+  bool isConstantDictionary_ = false;
+  bool isConstantUnicodeDictionaryBuilt_ = false;
+  bool isConstantAsciiDictionaryBuilt_ = false;
+
+  folly::F14FastMap<std::string, std::string> buildUnicodeDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    folly::F14FastMap<std::string, std::string> dictionary;
+    int i = 0;
+    int j = 0;
+    while (i < match.size()) {
+      std::string replaceChar;
+      // If match's character size is larger than replace's, the extra
+      // characters in match will be removed from input string.
+      if (j < replace.size()) {
+        int replaceCharLength = utf8proc_char_length(replace.data() + j);
+        replaceChar = std::string(replace.data() + j, replaceCharLength);
+        j += replaceCharLength;
+      }
+      int matchCharLength = utf8proc_char_length(match.data() + i);
+      std::string matchChar = std::string(match.data() + i, matchCharLength);
+      // Only considers the first occurrence of a character in match.
+      dictionary.emplace(matchChar, replaceChar);
+      i += matchCharLength;
+    }
+    return dictionary;
+  }
+
+  folly::F14FastMap<char, char> buildAsciiDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    folly::F14FastMap<char, char> dictionary;
+    int i = 0;
+    for (; i < std::min(match.size(), replace.size()); i++) {
+      char matchChar = *(match.data() + i);
+      char replaceChar = *(replace.data() + i);
+      // Only consider the first occurrence of a character in match.
+      dictionary.emplace(matchChar, replaceChar);
+    }
+    for (; i < match.size(); i++) {
+      char matchChar = *(match.data() + i);
+      dictionary.emplace(matchChar, '\0');
+    }
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* match,
+      const arg_type<Varchar>* replace) {
+    if (match != nullptr && replace != nullptr) {
+      isConstantDictionary_ = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace) {
+    if (!isConstantDictionary_ || !isConstantUnicodeDictionaryBuilt_) {
+      unicodeDictionary_ = buildUnicodeDictionary(match, replace);
+      if (!isConstantUnicodeDictionaryBuilt_) {
```

Comment:
Hi @mbasmanova, if dictionary is constant, it will be built only if it is not built yet, i.e., `isConstantUnicodeDictionaryBuilt_=false` here. Not building it for each row.

Reply:
Sorry, I meant to ask what happens if dictionary is not constant.

Reply:
If dictionary is not constant, which means `match` & `replace` are not constant, we have to build it for each row. Then, `isConstantUnicodeDictionaryBuilt_` means nothing.

Reply:
Do we have a test case that verifies that non-constant dictionary use case works properly?

Reply:
Hi @mbasmanova, I just added some tests for covering non-constant cases. Thanks!

Reply:
I got confused by the code setting isConstantUnicodeDictionaryBuilt_ to true even when dictionary is not constant. It happens to work, but I expect future readers to be confused as well. Consider changing the logic to 

```
if (isConstantDictionary_) {
   isConstantUnicodeDictionaryBuilt_ = true;
}
```

Reply:
Agree. It can be logically more clear. Just fixed. Thanks!